### ISSUE TITLE
YOLO

### ISSUE THREAD

**Issue by yushenxiang** [yushenxiang/MachineLearningTips#8]
摘要 ：
和之前的方法将物体检测视为分类的方法不同，yolo将物体检测作为回归问题进行处理；
一个单独的神经网络就可以将物体目标框和分类问题直接从全图中在一次测量中完成，实现端到端的检测。
yolo可以实现每秒45帧图片的测量，不会经常将背景误判为物体，但是定位精度略微差一些。
并且yolo可以较好的实现一般物体的检测，更加通用。

简介：
使用分类器的方法：
dpm使用滑动窗口的方法，rcnn使用region proposal生成可能的目标框，然后分类器进行分类任务，需要后处理来进行目标框重新修正，消除重复检测和重新打分。训练慢
yolo直接使用图像像素生成目标框坐标和物体分类。
yolo特点：
1）处理速度更快，实现实时监测
2）从图片全局进行判断，更少误判
3）能够对一般类型物体进行判断，更加通用
4）但精度方面还需改善，定位精度差

2统一的检测系统
1)将图片分成7行7列的栅格，如果物体中心在栅格内部，那么这栅格就用于对物体进行检测。
2)每个栅格会估计B（2）个目标框并进行打分。如果栅格里没有物体，那么confidence scores是0，否则confidence scores就是IOU.
每个目标框进行5个数值的估计：x, y, w,h,confidence.
The (x, y)代表物体中心相对栅格边界的坐标位置，w,h代表相对整个图片的宽度和高度，confidence.代表iou
3)每个栅格也会进行存在物体情况下的分类的判断。测试时通过将上述的结果相乘，就得到图片中每个目标框物体分类情况。
4)每个网格要预测B个bounding box还要预测C个categories。输出就是S x S x (5*B+C)的一个tensor
注意：class信息是针对每个网格的，confidence信息是针对每个bounding box的。
2。1结构设计
yolo:
根据googlenet改动，架构有20个卷积层和2个全链接层，没有使用iception ,使用了1x1的reduction layers和3x3的卷积层。

fast yolo:
9个卷积层，核尺寸更小，其他参数一样

2.2训练过程
1）在 ImageNet 1000-class competition dataset 进行于预训练，使用架构中的前20个卷积层，然后一个平均池化层和一个全链接层。然后将模型转化用于物体检测。
2）增加了4个卷积层和两个随机初始化的全链接层，检测需要更高精度的信息，因此输入图像是448x448.
3）最后一层用于物体分类和目标框坐标预测。最后一层使用线性激活函数，其他层使用leakyRelu.
4）模型的输出使用平方和优化，为避免给定位误差和分类误差同样权重，通过两个参数分别增加损失函数中定位误差的影响和降低损失函数中不存在物体的目标框的分类误差的影响。
同时，为了区分大目标框和小目标框对误差的影响程度，使用目标框的宽度和高度的平方根。
5）yolo的一个栅格会同时预测多个目标框。训练时，为了让一个目标框用于对应一个物体的预测，根据拥有最大IOU的目标框来用于对目标进行预计。

训练超参数：
1）Throughout training we use a batch size of 64, a momentum of 0.9 and a decay of 0.0005
2）Our learning rate schedule is as follows: For the first epochs we slowly raise the learning rate from 10 -3 to 10 -2。We continue training with 10- 2 for 75 epochs, then 10 -3 for 30 epochs, and finally 10 -4 for 30 epochs.
3）A dropout layer with rate = .5 after the first connected layer prevents co-adaptation between layer

2。3推理
由于近使用一个网络，yolo速度很快。但是由于可能存在大物体或栅格边界附近的物体会被多个栅格检测，因此使用最大值抑制的方法。

2.4yolo不足：
由于yolo里有很严格的空间限制，一个栅格只生成2个框，以及只判断一个物体，因此对于很小的物体，例如鸟，瓶子检测效果不好；
对测试图像中，同一类物体出现的新的不常见的长宽比和其他情况是。泛化能力偏弱。
由于定位精度不好，因此小目标框的影响较大；

3和其他检测方法对比：
1）DPM:
使用滑动窗口来进行目标检测。
提取静态特征，区域分类，为高得分区域预计目标框等是在隔离的系统里实现的。
yolo能在一个统一的网络中实现，更快更准确

2）rcnn和变种框架
使用region proposal代替滑动窗口进行检测。
rcnn使用selective search来生成目标框，卷积网络实现特征提取，SVM给每个目标框打分，一个线性模型重新调整目标框，最大值抑制方法消除重复检测。测试速度慢，一张图片40秒。
yolo严格的空间限制，降低每个栅格的目标框数目，一张图片98个目标框，而不是rcnn的2000歌目标框，降低重复检测，速度大大提升。

3）fast/faster rcnn'
通过共享计算和使用神经网络进行region proposal，放弃使用selective search，提升速度，但无法实现实时监测。

4）Deep MultiBox
使用神经网络来估计region of interest，而不使用selective search。能实现单一目标检测，不能实现一半物体检测，并且是一个复杂系统的一个子部分，需要额外的其他操作。

5）overfeat
使用神经网络来进行定位，并使用定位器来进行物体检测。overfeat仍使用滑动窗口方法，提升了定位精度，但检测精度没有提升。由于只能对局部图片信息进行处理，因此仍需要后处理。

6）MultiGrasp
yolo的根据栅格进行目标框的回归是给予MultiGrasp的原理实现的，不同是，MultiGrasp仅能判断某一个区域是否有物体，并不能判断物体尺寸，位置，以及物体种类和目标框坐标。

4实时监测系统对比：
1)Fast YOLO is the fastest object detection method on PASCAL;
2)30hz DPM actually produce a de- tection system that runs in real-time (30 frames per second or better);
3)Fast R-CNN  at 0.5 fps it is still far from real- time.
4) Zeiler- Fergus Faster R-CNN is only 2.5 times slower than YOLO but is also less accurate.

**Comment by yushenxiang** [comment]
http://blog.csdn.net/ben_ben_niao/article/details/52014285
yolo博客

**Comment by yushenxiang** [comment]
http://blog.csdn.net/majinlei121/article/details/53870433
目标检测方法系列——R-CNN, SPP, Fast R-CNN, Faster R-CNN, YOLO, SSD

**Comment by yushenxiang** [comment]
http://blog.csdn.net/ikerpeng/article/details/54316814
深度学习目标检测：RCNN，Fast，Faster，YOLO，SSD比较
